Temporary filter by infohash/save path/download path/comment/creator/files using existing filter name input

tristanleboss opened the pull request:
Implements "custom query" to filter all "string torrent data" from the actual name input filter.

Can use `(hash|save_path|download_path|comment|creator|filename):xxx`.

`filename` filter file names inside torrents. A great way to find the torrent containing the file you are searching for.

Even if colission with someone actually searching a torrent name containing a custom query should not happen often, a dropdown menu to toggle the function of the input field would probably be better but, at least, it works and all the filtering code can be reused. This can serve as a firt step and be upgraded later on.

Maybe adding a tooltip on the input field explaining these options can be a good idea.

tristanleboss commented:
It works!

tristanleboss commented:
I am stuck. If I paste "hash:0123456789abcdef0123456789abcdef01234567", it does the filtering. But, if I type this string letter by letter, it doesn't. It's because when typing h, a, s, h, :, it filters by name and so remove all torrents so at the time it detects its a hash search, there is no torrents left to filter. To avoid that, I thought calling `invalidateFilter` (through a public `disableNameFilter` method I created) when a hash search is detected and before applying it, would bring back all torrents by invalidating the filter set previously by `setFilterRegularExpression` but it doesn't seem so.

tristanleboss commented:
Ok, it finally works. I replaced `invalidateFilter` by `setFilterRegularExpression(u""_qs)` (an empty regexp so it match all torrents from the view).

tristanleboss commented:
Yes, it works perfectly!

glassez commented:
@tristanleboss 
Could you consider [qBittorrent Coding Guidelines](https://github.com/qbittorrent/qBittorrent/blob/master/CODING_GUIDELINES.md) and fix as much coding style issues as possible before we start reviewing?

glassez commented on src/gui/transferlistwidget.cpp:
```diff
@@ -1280,9 +1281,133 @@ void TransferListWidget::applyTrackerFilter(const QSet<BitTorrent::TorrentID> &t
+        }
+        else
+        {
+            this->disablePreviousCustomQueryFilter();
```
In C++, we don't explicitly use `this` if it is not required.

tristanleboss commented:
@glassez I removed the not-needed this and read why they are not needed (I didn't remember from my courses). I fixed few things.

I am setting up a linux VM with the complete build tool chain because I failed to load the Qt header files in my Windows VSCode and sometime comited stupid code.

tristanleboss commented:
@Chocobo1 @glassez Is it possible to have your opinion on this PR. Indeed, I have completed a UI for this an

![image](https://user-images.githubusercontent.com/1431508/195883137-f9d1249b-f084-43ab-9e0e-8fdd35dde03e.png)

My big questions are:
- should I continue working on this? Would you consider merging such a feature?
- can I reuse the filtering code of this PR? Is it done the way you want it to?

Chocobo1 commented:
>@Chocobo1 @glassez Is it possible to have your opinion on this PR. Indeed, I have completed a UI for it (not yet merged with this PR code) and I would love for such a feature to land in 4.5.0.

@qbittorrent/bug-handlers any opinions?

---

I can only answer a subset.
>should I continue working on this? 

The UI looks good. Although I don't recall many users requested this feature, it should still be useful.

>Would you consider merging such a feature?

If you meant the UI: from a user perspective, I would want it. As a developer, making a decision now is too early...

thalieht commented:
>https://github.com/orgs/qbittorrent/teams/bug-handlers any opinions?

I barely use the name filter as is and it is perfectly adequate for me. Looks good though. I see no reason not to merge it.

glassez commented:
> Indeed, I have completed a UI for it

UI looks good to me. And it is much better than typing filter type manually. But in case of having UI for selecting the filtered field, the implementation of applying the filter could be more optimal by separating the filter pattern and filtering field (name, save path, hash, etc.).

stalkerok commented:
It is a pity that the author abandoned this PR. @glassez, all hope is on you, especially since users periodically request this function https://github.com/qbittorrent/qBittorrent/issues/19627

stalkerok commented:
@tristanleboss, could you please finish working on these features?

estatistics commented:
How we can merge /add this tool into the existing qbittorent? can you explain?

estatistics commented:
How i can see what files where edited, to compared them with the last version I have and merge the changes accordingly?

estatistics commented on src/gui/transferlistwidget.cpp:
1) const auto has now (22sep2023) double parenthesis before QT. 
2) auto filedialog, auto dialog, auto menu, auto trackingdialog,  have "*" before its name. 
3) paths.insert changed     to `paths.insert(!contentPath.isEmpty() ? contentPath : torrent->savePath());`

This file must be reviewed again.

luzpaz commented:
You can create your own fork of qBittorrent. Then you download the https://github.com/tristanleboss/qBittorrent/tree/patch-2 branch and use git checkout to view it. 
Using git you can run a git rebase and then fix the merge conflicts.
 
These are the files that the original PR made changes to:
![Screenshot_20230923_232635](https://github.com/qbittorrent/qBittorrent/assets/4140247/afb6b8e1-98f6-41d3-b393-3db2423a1ada)

estatistics commented:
I tried to compile from source qbitorrent in my system with these changes, but it failed on other level. QT6.5 is not yet available to my system. so i must may download an older version of qbit to apply these changes to see if this works or to wait for qt6.5 to be available to my system.

stalkerok commented:
It was a promising feature, but the author seems to have disappeared with the ends. @glassez @Chocobo1 do you have any plans to implement something similar?
